crates/aptos-crypto/src/poseidon_bn254/mod.rs

```diff
@@ -0,0 +1,147 @@
+// Copyright (c) Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+//! Implements the Poseidon hash function for BN-254, which hashes $\le$ 16 field elements and
+//! produces a single field element as output.
+mod alt_fr;
+mod constants;
+pub mod keyless;
+
+use crate::poseidon_bn254::constants::*;
+use anyhow::bail;
+use keyless::{
+    pack_bytes_to_scalars, pad_and_pack_bytes_to_scalars_no_len,
+    pad_and_pack_bytes_to_scalars_with_len,
+};
+use neptune::poseidon::{HashMode::OptimizedStatic, Poseidon};
+
+/// The maximum number of input scalars that can be hashed using the Poseidon-BN254 hash function
+/// exposed in `hash_scalars`.
+pub const MAX_NUM_INPUT_SCALARS: usize = 16;
+
+/// Macro for Poseidon-BN254-hashing a vector of scalars.
+macro_rules! neptune_hash {
+    ($elems:expr, $constants:expr) => {{
+        let mut hasher = Poseidon::new(&$constants);
+        hasher.reset();
+        for elem in $elems.into_iter() {
+            hasher.input(elem.into()).expect("Too many inputs");
+        }
+        hasher.hash_in_mode(OptimizedStatic);
+        hasher.elements[0]
+    }};
+}
+
+/// Given an array of up to `MAX_NUM_INPUT_SCALARS` field elements (in the BN254 scalar field), hashes
+/// them using Poseidon-BN254 into a single field element.
+pub fn hash_scalars(inputs: Vec<ark_bn254::Fr>) -> anyhow::Result<ark_bn254::Fr> {
+    if inputs.is_empty() || inputs.len() > MAX_NUM_INPUT_SCALARS {
+        bail!(
+            "Poseidon-BN254 needs > 0 and <= {} inputs, but was called with {} inputs",
+            MAX_NUM_INPUT_SCALARS,
+            inputs.len()
+        );
+    }
+
+    let result = match inputs.len() {
```

Comment:
Nit that's not really worth changing but you could probably do this with a macro

Reply:
I tried writing another macro that calls this one, but Rust's macro system does not support that.